### PR TITLE
[Feature] [add/update]DocumentsInBatches

### LINKED ISSUE #95: Add methods to automatically add/update documents in batches
We need methods to add and update the documents of MeiliSearch in batch instead of letting the users call `addDocuments` in a loop.

- [ ] Add `addDocumentsInBatches`
```
addDocumentsInBatches(documents, batchSize = 1000, primaryKey = nil)
```
`batchSize` and `primaryKey` are optional parameters

- [ ] Add `updateDocumentsInBatches`
```
updateDocumentsInBatches(documents, batchSize = 1000, primaryKey = nil)
```
`batchSize` and `primaryKey` are optional parameters

- [ ] Add tests

Example: https://github.com/meilisearch/meilisearch-python/pull/260

⚠️ This issue is generated, it means `addDocumentsInBatches` and `updateDocumentsInBatches` might be named differently in this package. Keep the already existing way of naming in this package to stay idiomatic with the language and the repository.
Also, if you are a maintainer, feel free to add any clarification and instruction about this issue related to the specific languages of the repo.

Related to https://github.com/meilisearch/integration-guides/issues/106

Sorry if this is already partially/completely implemented, feel free to let me know about the state of this issue in the repo.

### PR DESCRIPTION
**Pull request by ahmednfwela**
Fixes #95
- [x] Tests added

### CODE REVIEW

**Comment by ahmednfwela** [comment]
cc @brunoocasali

File: lib/src/index.dart [hunk 1]
```diff
@@ -60,13 +60,29 @@ abstract class MeiliSearchIndex {
     String? primaryKey,
   });
 
+  /// Add a list of documents in batches of size [batchSize] by given [documents] and optional [primaryKey] parameter.
+  /// If index is not exists tries to create a new index and adds documents.
```
Comment (brunoocasali) [inline lib/src/index.dart:64]:
```suggestion
  /// Add a list of documents in batches of size [batchSize] by giving [documents] and optional [primaryKey] parameter.
  /// If the index does not exist try to create a new index and add documents.
```
```

File: test/documents_test.dart [hunk 2]
```diff
@@ -37,6 +52,31 @@ void main() {
       expect(doc?['title'], equals('The Hobbit 2'));
     });
 
+    test('Update documents in batches', () async {
+      const batchSize = 10;
```
Comment (brunoocasali) [inline test/documents_test.dart:56]: This is an excellent recommendation I like to share whenever is possible: http://blog.plataformatec.com.br/2014/04/improve-your-test-readability-using-the-xunit-structure/

By using this pattern your test structure will look way more readable!

File: test/utils/wait_for.dart [hunk 3]
```diff
@@ -22,6 +23,38 @@ extension TaskWaiter on Task {
...
+            .whereNotNull()
+            .toList();
+      }
+      await Future<void>.delayed(interval);
+    }
+
+    throw Exception('The tasks $originalUids timed out.');
+  }
+}
+
```
Comment (brunoocasali) [inline test/utils/wait_for.dart:57]:
```suggestion
extension TaskWaiterForLists on Iterable<Task> {
  Future<List<Task>> waitFor({
    required MeiliSearchClient client,
    Duration timeout = const Duration(seconds: 20),
    Duration interval = const Duration(milliseconds: 50),
  }) async {
    final endingTime = DateTime.now().add(timeout);
    final originalUids = toList();
    final remainingUids = map((e) => e.uid).whereNotNull().toList();
    final completedTasks = <int, Task>{};
    final statuses = ['enqueued', 'processing'];

    while (DateTime.now().isBefore(endingTime)) {
      var query = TasksQuery(uids: remainingUids);

      final tasks = await client.getTasks(params: query).results;
      final completed = tasks.where((e) => statuses.contains(e.status));

      completedTasks.addEntries(completed.map((e) => MapEntry(e.uid!, e)));
      remainingUids.removeWhere((e) => completedTasks.containsKey(e));

      if (remainingUids.isEmpty) {
        return originalUids
            .map((e) => completedTasks[e.uid])
            .whereNotNull()
            .toList();
      }

      await Future<void>.delayed(interval);
    }

    throw Exception('The tasks $originalUids timed out.');
  }
}
```

File: test/documents_test.dart (hunk 2 above)
Comment (ahmednfwela) [reply test/documents_test.dart:57]: Thanks for this tip, definitely will use it from now on!

**Review by brunoocasali — APPROVED** [review]
bors merge

**Pull request merged**